File: 백준/14503.py
import sys
from collections import deque

sys.stdin = open('data.txt', 'r')

N, M = map(int, sys.stdin.readline().rstrip().split(" "))
r, c, d = map(int, sys.stdin.readline().rstrip().split(" "))

room = []
for i in range(N):
    room.append(list(map(int, sys.stdin.readline().rstrip().split(" "))))

# 0 1 2 3 : 북 동 남 서
cur_direction = deque([(i + d) % 4 for i in range(4, 0, -1)])


def turn_on(r, c):
    global answer

    if room[r][c] == 0:
        room[r][c] = 2
        answer += 1

    count = 0
    if (0 <= r - 1 < N) and (room[r - 1][c]) == 0:
        count += 1
    if (0 <= r + 1 < N) and (room[r + 1][c]) == 0:
        count += 1
    if (0 <= c + 1 < M) and (room[r][c + 1]) == 0:
        count += 1
    if (0 <= c - 1 < M) and (room[r][c - 1]) == 0:
        count += 1

    if count == 0:  # 빈칸이 없는 경우
        if cur_direction[0] == 0:  # 아래로
            if 0 < r + 1 < N - 1:
                if room[r + 1][c] == 1:
                    return
                else:
                    turn_on(r + 1, c)
        elif cur_direction[0] == 1:  # 왼쪽으로
            if 0 < c - 1 < M - 1:
                if room[r][c - 1] == 1:
                    return
                else:
                    turn_on(r, c - 1)
        elif cur_direction[0] == 2:  # 위로
            if 0 < r - 1 < N - 1:
                if room[r - 1][c] == 1:
                    return
                else:
                    turn_on(r - 1, c)
        elif cur_direction[0] == 3:  # 오른쪽으로
            if 0 < c + 1 < M - 1:
                if room[r][c + 1] == 1:
                    return
                else:
                    turn_on(r, c + 1)
    else:  # 빈칸이 있는 경우
        for _ in range(4):
            cur_direction.rotate(-1)
            if cur_direction[0] == 0:  # 위
                if (0 <= r - 1 < N) and (room[r - 1][c] == 0):
                    turn_on(r - 1, c)
                    break
            elif cur_direction[0] == 1:  # 오른쪽
                if (0 <= c + 1 < M) and (room[r][c + 1] == 0):
                    turn_on(r, c + 1)
                    break
            elif cur_direction[0] == 2:  # 아래
                if (0 <= r + 1 < N) and (room[r + 1][c] == 0):
                    turn_on(r + 1, c)
                    break
            elif cur_direction[0] == 3:  # 왼쪽
                if (0 <= c - 1 < M) and (room[r][c - 1] == 0):
                    turn_on(r, c - 1)
                    break


answer = 0
turn_on(r, c)
print(answer)

File: 백준/14503.js
const path =
  process.platform === "linux" ? "/dev/stdin" : __dirname + "\\data.txt";
const input = require("fs").readFileSync(path).toString().trim().split(/\r?\n/);

let lineIdx = 0;
const [N, M] = input[lineIdx++].split(" ").map(Number);
const robot = input[lineIdx++].split(" ").map(Number);
const room = [];
for (let i = 0; i < N; i++) {
  room.push(input[lineIdx++].split(" ").map(Number));
}

function solution(N, M, robot, room) {
  const isVisited = Array.from(new Array(N), () => new Array(M).fill(false));
  const vectors = [
    [-1, 0],
    [0, 1],
    [1, 0],
    [0, -1],
  ];
  let [curY, curX, curV] = robot;
  let answer = 0;

  while (true) {
    if (!isVisited[curY][curX]) {
      // 청소되지 않은 경우
      isVisited[curY][curX] = true;
      answer++;
    }

    let isSearch = false;
    for (let [dy, dx] of vectors) {
      let [serachY, searchX] = [curY + dy, curX + dx];

      if (!(0 <= serachY && serachY < N && 0 <= searchX && searchX < M))
        continue;

      if (!isVisited[serachY][searchX] && room[serachY][searchX] === 0) {
        // 주변 4칸중 청소되지 않은 빈칸이 있는 경우
        curV = (curV - 1 >= 0 ? curV - 1 : 4 + (curV - 1)) % 4; // 반시계 방향 회전
        let [nextY, nextX] = [curY + vectors[curV][0], curX + vectors[curV][1]];

        if (!(0 <= nextY && nextY < N && 0 <= nextX && nextX < M)) continue;

        if (!isVisited[nextY][nextX] && room[nextY][nextX] === 0) {
          // 바라보는 방향 기준으로 앞쪽칸이 청소되지 않은 빈칸이면
          [curY, curX] = [nextY, nextX];
        }
        isSearch = true;
        break;
      }
    }

    if (isSearch) continue;

    let backV = (curV + 2) % 4;
    let [backY, backX] = [curY + vectors[backV][0], curX + vectors[backV][1]];

    if (!(0 <= backY && backY < N && 0 <= backX && backX < M)) continue;

    if (room[backY][backX] === 0) {
      [curY, curX] = [backY, backX];
    } else {
      console.log(answer);
      return;
    }
  }
}

solution(N, M, robot, room);
